uaclient/actions.py

```diff
@@ -117,3 +119,62 @@ def status(
         ret = 0
 
     return status, ret
+
+
+def get_cloud_instance(
+    cfg: config.UAConfig,
+) -> Optional[AutoAttachCloudInstance]:
+    disable_auto_attach = util.is_config_value_true(
+        config=cfg.cfg, path_to_value="features.disable_auto_attach"
+    )
+    if disable_auto_attach:
+        msg = "Skipping auto-attach. Config disable_auto_attach is set."
+        logging.debug(msg)
+        print(msg)
+        return None
+
+    instance = None  # type: Optional[AutoAttachCloudInstance]
+    try:
+        instance = identity.cloud_instance_factory()
+    except exceptions.CloudFactoryError as e:
+        if cfg.is_attached:
+            # We are attached on non-Pro Image, just report already attached
+            raise exceptions.AlreadyAttachedError(cfg)
+        if isinstance(e, exceptions.CloudFactoryNoCloudError):
+            raise exceptions.UserFacingError(
+                messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+            )
+        if isinstance(e, exceptions.CloudFactoryNonViableCloudError):
+            raise exceptions.UserFacingError(messages.UNSUPPORTED_AUTO_ATTACH)
+        if isinstance(e, exceptions.CloudFactoryUnsupportedCloudError):
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH_CLOUD_TYPE.format(
+                    cloud_type=e.cloud_type
+                )
+            )
+        # we shouldn't get here, but this is a reasonable default just in case
+        raise exceptions.UserFacingError(
+            messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+        )
+
+    if not instance:
+        # we shouldn't get here, but this is a reasonable default just in case
+        raise exceptions.UserFacingError(
+            messages.UNABLE_TO_DETERMINE_CLOUD_TYPE
+        )
+    return instance
+
+
+def detach_before_auto_attach(cfg: config.UAConfig):
+    from uaclient.cli import _detach
```

Comment:
This could be problematic, if we are using the `cli` module, we might be leaking `event.info`calls for the api.
Additionally, we could also prompt users here, which is not ideal.

I think we should refactor this to allow us to be decouple from any `cli` function

Reply:
@CalvoM I just merged https://github.com/canonical/ubuntu-advantage-client/pull/2176 so you no longer need the `detach_before_auto_attach` function :tada: